dragonfly/block/sponge.go

```diff
@@ -0,0 +1,94 @@
+package block
+
+import (
+	"github.com/df-mc/dragonfly/dragonfly/item"
+	"github.com/df-mc/dragonfly/dragonfly/world"
+	"github.com/go-gl/mathgl/mgl64"
+)
+
+// Sponge is a block that can be used to remove water around itself when placed, turning into a wet sponge in the
+// process.
+type Sponge struct {
+	// Wet specifies whether the dry or the wet variant of the block is used.
+	Wet bool
+}
+
+// BreakInfo ...
+func (s Sponge) BreakInfo() BreakInfo {
+	return BreakInfo{
+		Hardness: 0.6,
+		Drops:    simpleDrops(item.NewStack(s, 1)),
+	}
+}
+
+// EncodeItem ...
+func (s Sponge) EncodeItem() (id int32, meta int16) {
+	if s.Wet {
+		meta = 1
+	}
+
+	return 19, meta
+}
+
+// EncodeBlock ...
+func (s Sponge) EncodeBlock() (name string, properties map[string]interface{}) {
+	if s.Wet {
+		return "minecraft:sponge", map[string]interface{}{"sponge_type": "wet"}
+	}
+	return "minecraft:sponge", map[string]interface{}{"sponge_type": "dry"}
+}
+
+// UseOnBlock ...
+func (s Sponge) UseOnBlock(pos world.BlockPos, face world.Face, v mgl64.Vec3, w *world.World, user item.User, ctx *item.UseContext) (used bool) {
+	// the sponge is already wet so it cannot consume any more water.
+	if !s.Wet {
```

Comment:
On bedrock edition it seems to work a little different. Sponges only absorb water if they are directly adjacent to a water block. They can also absorb water if water flows adjacent to them after having been placed and if they're still dry. You can use the NeighbourUpdateTick method for that.

Reply:
You might want to pull the absorbing behaviour out of this method into a new method and call it on NeighbourUpdateTick and on placement, while checking if there is water directly adjacent to it.